torch/csrc/jit/script/compiler.cpp

```diff
@@ -559,25 +593,95 @@ struct to_ir {
+      }
+    }
+
+    if (num_starred > 1) {
+      throw ErrorReport(r)
+          << "Only one starred expression is allowed on the lhs.";
+    }
+
+    if (num_starred > 0 && num_normal_assign == 0) {
+      throw ErrorReport(r) << "A Starred expression may only appear on the "
+                              << "lhs within the presence of another non-starred"
+                              << " expression.";
+    }
+
+    return num_starred;
+  }
+
+  std::vector<SugaredValuePtr> createSugaredValuesFromValues(const std::vector<Value*> values) {
+    std::vector<SugaredValuePtr> sugared_outputs;
+    sugared_outputs.reserve(values.size());
+    for (Value* output : values) {
+      sugared_outputs.emplace_back(std::make_shared<SimpleValue>(output));
+    }
+    return sugared_outputs;
+  }
+
   std::vector<Value*> emitAssignment(const Assign& stmt) {
     std::vector<Value*> outputs{stmt.lhs().size()};
+    size_t num_starred_unpack = calcNumStarredUnpack(stmt.lhs(), stmt.range());
     if (stmt.reduction() != '=') {
       if (stmt.lhs().size() != 1) {
         throw ErrorReport(stmt)
-            << "reductions are only allow when there is a single variable "
+            << "reductions are only allowed when there is a single variable "
             << "on the left-hand side.";
       }
-      Ident lhs = stmt.lhs()[0];
+      Ident lhs = Var(stmt.lhs()[0]).name();
       Expr expr = BinOp::create(stmt.range(), stmt.reduction(),
                                 Var::create(lhs.range(), lhs), stmt.rhs());
-      outputs = emitExpr(expr, 1);
+      outputs = emitExpr(expr, {1, false});
     } else {
-      outputs = emitExpr(stmt.rhs(), stmt.lhs().size());
+      CallsiteDescriptor cd{stmt.lhs().size(), !!num_starred_unpack || stmt.lhs().size() == 1};
+      outputs =
+          emitExpr(stmt.rhs(), cd);
     }
-    int i = 0;
-    for (auto ident : stmt.lhs()) {
-      environment_stack->setVar(Ident(ident).name(), outputs.at(i));
-      i++;
+    if (stmt.lhs().size() == 1 && outputs.size() != 1) {
+      // Pack up a tuple sugared value
+      SugaredValuePtr tup = std::make_shared<TupleValue>(createSugaredValuesFromValues(outputs));
+      if (stmt.lhs()[0].kind() != TK_VAR) {
+        throw ErrorReport(stmt.lhs()[0]) << "Cannot pack a tuple into a non-variable.";
+      }
+      environment_stack->setSugaredVar(Var(stmt.lhs()[0]).name().name(), tup);
+    } else {
+      int i = 0;
+      for (auto assignee : stmt.lhs()) {
+        if (assignee.kind() == TK_VAR) {
+          environment_stack->setVar(Var(assignee).name().name(), outputs.at(i));
+          i++;
+        } else if (assignee.kind() == TK_STARRED) {
+          auto var = Starred(assignee).expr();
+          if (var.kind() != TK_VAR) {
+            throw ErrorReport(var) << "Cannot pack a tuple into a non-variable.";
+          }
+          std::vector<Value*> starred_slice(
+              outputs.begin() + i, outputs.begin() + i + num_starred_unpack);
```

Comment:
Is this correct? It seems to me like `num_starred_unpack` is always 1 in this case?